plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -623,4 +640,100 @@ private static void validateOrcBloomFilterColumns(ConnectorTableMetadata tableMe
+            }
+            if (latestMetadataLocations.size() > 1) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("More than one latest metadata file found at location: %s, latest metadata files are %s",
+                        metadataDirectoryLocation, latestMetadataLocations));
+            }
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed checking table's location: " + location, e);
+        }
+        return Optional.of(getOnlyElement(latestMetadataLocations.iterator()));
+    }
+
+    /**
+     * Get the latest metadata file location present in location if metadataFileName is not provided, otherwise
+     * form the metadata file location using location and metadataFileName
+     *
+     * @param fileSystem TrinoFileSystem
+     * @param location Table location
+     * @param metadataFileName metadata file name
+     * @return full qualified metadata file location
+     */
+    public static Optional<String> getMetadataLocation(TrinoFileSystem fileSystem, String location, Optional<String> metadataFileName)
+    {
+        if (metadataFileName.isEmpty()) {
+            return getLatestMetadataLocation(fileSystem, location);
+        }
+        else {
+            return Optional.of(format("%s/%s/%s", stripTrailingSlash(location), METADATA_FOLDER_NAME, metadataFileName.get()));
+        }
+    }
+
+    private static OptionalInt parseVersion(String metadataLocation)
+    {
+        // TODO may be make the method at one place. same as AbstractIcebergTableOperations#parseVersion
+        int versionStart = metadataLocation.lastIndexOf('/') + 1; // if '/' isn't found, this will be 0
+        int versionEnd = metadataLocation.indexOf('-', versionStart);
+        try {
+            return OptionalInt.of(parseInt(metadataLocation.substring(versionStart, versionEnd)));
+        }
+        catch (NumberFormatException | IndexOutOfBoundsException e) {
+            return OptionalInt.empty();
+        }
+    }
+
+    /**
+     * Read metadata from provided metadata file location. This API should be called for valid metadata location.
+     *
+     * @param fileSystem TrinoFileSystem
+     * @param metadataLocation Metadata file location
+     * @return TableMetadata
+     */
+    public static TableMetadata readMetadata(TrinoFileSystem fileSystem, String metadataLocation)
+            throws TrinoException
+    {
+        FileIO fileIO = fileSystem.toFileIo();
+        InputFile metadataFile = fileIO.newInputFile(metadataLocation);
+        return TableMetadataParser.read(fileIO, metadataFile);
+    }
+
+    public static void validateLocation(TrinoFileSystem fileSystem, String location)
+    {
+        try {
+            if (!fileSystem.newInputFile(location).exists()) {
+                throw new TrinoException(ICEBERG_FILE_NOT_FOUND, format("Location %s does not exist", location));
```

Comment:
for user-provided location this should be a USER_ERROR category
for something we read from metastore, ICEBERG_INVALID_METADATA would be appropriate
for something we calculated based on file system state, we don't need to validate

Reply:
Used the error code accordingly. Let me know if any other error code needs to be modify.